keycloak-iam/src/main/java/com/valtimo/keycloak/security/jwt/authentication/KeycloakTenantTokenAuthenticator.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2015-2020 Ritense BV, the Netherlands.
+ *
+ * Licensed under EUPL, Version 1.2 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.valtimo.keycloak.security.jwt.authentication;
+
+import com.ritense.tenancy.TenantResolver;
+import com.ritense.tenancy.jpa.TenantAwareListener;
+import com.ritense.tenancy.web.DelegatingTenantAuthenticationToken;
+import com.ritense.valtimo.contract.security.jwt.TokenAuthenticator;
+import io.jsonwebtoken.Claims;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.User;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.ritense.valtimo.contract.authentication.AuthoritiesConstants.USER;
+import static com.ritense.valtimo.contract.security.jwt.JwtConstants.*;
+import static java.util.Objects.requireNonNull;
+
+public class KeycloakTenantTokenAuthenticator extends TokenAuthenticator {
```

Comment:
This class is almost the same are the original. I don't know if this is the best approach. Meaning you could argue that on line 89 where the authentication type is chosen this is done for both scenarios.